Fix errors in rotation function in OSMPDummySensor

ClemensLinnhoff opened the pull request:
#### Reference to a related issue in the repository
Fixes #94

#### Add a description
- Added a transpose function for the rotation matrix, as it is a transformation from a global coordinate system to a local system and therefore, the inverse needs to be used
- Changed using the objects orientation to using the ego orientation for calculating the relative position of an object with respect to the ego vehicle.

#### Check the checklist

- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the [documentation](https://github.com/OpenSimulationInterface/osi-documentation) for osi-sensor-model-packaging.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works: [Failed unit test](https://github.com/openMSL/sl-1-0-sensor-model-repository-template/actions/runs/4414328802/jobs/7735892106), [Passing unit test](https://github.com/openMSL/sl-1-0-sensor-model-repository-template/actions/runs/4414405272/jobs/7736070483)
- [x] New and existing unit tests / Github Actions pass locally with my changes.

ClemensLinnhoff committed:
Added transform function for rotation matrix and changed object orientation to ego orientation

Signed-off-by: ClemensLinnhoff <clemens.linnhoff@persival.de>

PhRosenberger reviewed (commented):
As this occured during unit testing and is now solved, it should be merged asap.

thomassedlmayer committed:
Remove rotatePoint in DummySource

Signed-off-by: Thomas Sedlmayer <tsedlmayer@pmsfit.de>

thomassedlmayer committed:
Fix OSMPDummySensor's coordinate transformations

- Change transformation matrix to XYZ rotation order acc. to OSI3 def.
- Consider center rear axle for transformation
- Consider virt. sensor mounting position/orientation for transformation
- Copy/translate object orientations from SensorView to SensorData
- Copy virtual sensor mounting position from SensorView to SensorData

Signed-off-by: Thomas Sedlmayer <tsedlmayer@pmsfit.de>

thomassedlmayer commented:
Thanks for noticing/fixing the bug! The fix works with our tools.

Though, I thought we would maybe want to use the actual roll-pitch-yaw rotation matrix instead of using the yaw-pitch-roll matrix and then transpose? I made a proposal in my commit. What do you think?

I also added the transformation to vehicle coordinate system (center rear axle) and virtual sensor coordinate system as this was missing anyways. And I removed the rotatePoint function from the DummySource as this probably was a copy/paste mistake.

@ClemensLinnhoff Could you check if this works for you?

ClemensLinnhoff commented on examples/OSMPDummySensor/OSMPDummySensor.cpp:
```diff
@@ -246,15 +258,60 @@ void rotatePoint(double x, double y, double z,double yaw,double pitch,double rol
+
+    /* rotate by negative ego vehicle orientation */
+    rotatePointXYZ(rx, ry, rz,
+        -ego_yaw, -ego_pitch, -ego_roll,
```
I think it is a bit unintuitive, that you have to call the function with negative angles.

ClemensLinnhoff commented on examples/OSMPDummySensor/OSMPDummySensor.cpp:
```diff
@@ -246,15 +258,60 @@ void rotatePoint(double x, double y, double z,double yaw,double pitch,double rol
+
+    /* rotate by negative virtual/physical sensor mounting orientation */
+    rotatePointXYZ(rx, ry, rz,
+        -mounting_position_yaw, -mounting_position_pitch, -mounting_position_roll,
```
Same as above

ClemensLinnhoff commented on examples/OSMPDummySensor/OSMPDummySensor.cpp:
```diff
@@ -287,20 +350,27 @@ fmi2Status COSMPDummySensor::doCalc(fmi2Real currentCommunicationPoint, fmi2Real
-                    double trans_z = veh.base().position().z()-ego_z;
-                    double rel_x,rel_y,rel_z;
-                    rotatePoint(trans_x,trans_y,trans_z,veh.base().orientation().yaw(),veh.base().orientation().pitch(),veh.base().orientation().roll(),rel_x,rel_y,rel_z);
+                    double rel_x, rel_y, rel_z;
```
We can skip this, because the position is not realtive here anymore. I like, that you put the entire transformation into a separate function. So it can just directly be called with veh.base().position(). It would probably also be easier not to separate between x,y,z in the function call and just pass an osi::vector3d.

ClemensLinnhoff replied:
After we talked I understand the approach not to use OSI types for internal functions. So this is fine with me.

ClemensLinnhoff reviewed (commented):
Apart from the individual comments in the code, I would acutally prefer to use the inverse of the rotation matrix instead of reversing the rotation order and the angles, because it would then be alligned with the mathematical definitions in the [documentation](https://opensimulationinterface.github.io/osi-documentation/#_coordinate_transformations).

ClemensLinnhoff commented:
As I said above, for me it is easier to understand the coordinate transformtion by having only one rotation matrix and using the inverse depending on the transformation direction. But I can understand, that people might also find it easier doing it by reversing the rotation direction and the angle directions. So it would be fine for me to use your approach @thomassedlmayer 

I tested the current implementation with some unit tests and could not find an error, so the transformation (incl. translation and rotation) to vehicle and to the sensor seem to work.

ClemensLinnhoff commented on examples/OSMPDummySensor/OSMPDummySensor.cpp:
```diff
@@ -236,7 +236,19 @@ fmi2Status COSMPDummySensor::doExitInitializationMode()
 }
 
-void rotatePoint(double x, double y, double z,double yaw,double pitch,double roll,double &rx,double &ry,double &rz)
+void transposeRotationMatrix(double matrix_in[3][3], double matrix_trans[3][3]) {
```
This function can be removed, as it is no longer needed.

ClemensLinnhoff commented:
After removing the transpose function as noted above, the PR can be merged from my point of view.

thomassedlmayer committed:
Minor clean-up

Signed-off-by: Thomas Sedlmayer <tsedlmayer@pmsfit.de>

pmai commented:
CCB 2023-05-22: Merge as-is.
